### DISCUSSION TITLE
ui.notify restets connection / shows after function is completed

### DISCUSSION THREAD

**Discussion by Nikodemski2002** [zauberzeug/nicegui#937] [Q&A] [answered]
### Question

Hi everyone. I have a problem with the implementation of ui.notify('Loading ...', type='ongoing') (basic ui.notify('Loading ...') also), namely when I try to show notification before the start of the function, it doesn't show after clicking the button that initiates the function **but after the function is completed together with the 'Simulation completed' notification**. Also (wich is more strange) sometimes the app restarts the connection if there is any ui.* element before function calc() call.

The non-working code example:
```py
def startSim():
    ui.notify('Loading...')
    df = calc()
    ui.notify('Simulation completed', type='positive')

ui.button('Start simulation', on_click=lambda: startSim())
```

Example error that shows:
```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\server.py", line 248, in handler
    await self.close()
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\protocol.py", line 766, in close
    await self.write_close_frame(Close(code, reason))
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\protocol.py", line 1232, in write_close_frame
    await self.write_frame(True, OP_CLOSE, data, _state=State.CLOSING)
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\protocol.py", line 1205, in write_frame
    await self.drain()
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\protocol.py", line 1194, in drain
    await self.ensure_open()
  File "C:\Users\nikon\PycharmProjects\4GEN_ANALYTICS\venv\lib\site-packages\websockets\legacy\protocol.py", line 935, in ensure_open
    raise self.connection_closed_exc()
websockets.exceptions.ConnectionClosedError: sent 1000 (OK); no close frame received
```

For example (shown below), if I use the ui.notify() only after the calc(0 function call, everything works perfectly.
The working code example:
```py
def startSim():
    df = calc()
    ui.notify('Simulation completed', type='positive')

ui.button('Start simulation', on_click=lambda: startSim())
```

Best regards
Nikodem

**Comment by falkoschindler** [comment, marked as answer]
As catamountjack wrote [on Discord]( https://discord.com/channels/1089836369431498784/1108497400760377455/1108596262665723954) just recently:

> [...] The problem is that [`calc()`] is a blocking call so none of the ui events are getting updated until the function is complete.  NiceGUI is built on FastAPI which is uses an asynchronous framework to allow ui updates while the program runs.  If you convert your function into an async function and use [`await calc()`], events should appear as you expect. [...]

I guess that `calc()` is blocking so long that the client disconnects. See https://github.com/zauberzeug/nicegui/discussions/590 on how to run long computations with `run_in_executor`.

**Reply by Nikodemski2002** [reply]
Hi. @falkoschindler thanks for the answer. I've used an additional calling function and removed lambda from ui.button. I've also changed the ui.notify('Loading', type='ongoing') for ui.spinner, cause as I see, for now, there is no way to close ui.notify(type='ongoing').
Here's working code example:
```py
def passSimParams():
    result = calc()
    return result

async def startSim():
    spinner = ui.spinner(size='lg').style('margin-left: auto; margin-right: auto;')
    result = await asyncio.get_event_loop().run_in_executor(None, passSimParams)
    spinner.set_visibility(False)
    ui.notify('Simulation completed', type='positive')

ui.button('Start simulation', on_click=startSim)
```
Best regards
Nikodem
